dvc/command/metrics.py

```diff
@@ -12,33 +12,57 @@
 
 
 def _show_metrics(
-    metrics, all_branches=False, all_tags=False, all_commits=False
+    metrics,
+    all_branches=False,
+    all_tags=False,
+    all_commits=False,
+    precision=None,
 ):
-    from dvc.utils.diff import format_dict
+    from dvc.utils.diff import format_dict, table
     from dvc.utils.flatten import flatten
 
     # When `metrics` contains a `None` key, it means that some files
     # specified as `targets` in `repo.metrics.show` didn't contain any metrics.
     missing = metrics.pop(None, None)
+    with_rev = any([all_branches, all_tags, all_commits])
+    header_set = set()
+    rows = []
 
-    lines = []
-    for branch, val in metrics.items():
-        if all_branches or all_tags or all_commits:
-            lines.append(f"{branch}:")
+    def _round(val):
+        if isinstance(val, float):
+            return round(val, precision)
+        return val
 
+    for _branch, val in metrics.items():
+        for _fname, metric in val.items():
+            if not isinstance(metric, dict):
+                header_set.add("")
+                continue
+            for key, _val in flatten(format_dict(metric)).items():
+                header_set.add(key)
+    header = sorted(header_set)
+    for branch, val in metrics.items():
         for fname, metric in val.items():
+            row = []
+            if with_rev:
+                row.append(branch)
+            row.append(fname)
             if not isinstance(metric, dict):
-                lines.append("\t{}: {}".format(fname, str(metric)))
+                row.append(str(metric))
+                rows.append(row)
                 continue
+            flattened_val = flatten(format_dict(metric))
 
-            lines.append(f"\t{fname}:")
-            for key, value in flatten(format_dict(metric)).items():
-                lines.append(f"\t\t{key}: {value}")
+            for i in header:
+                row.append(_round(flattened_val.get(i)))
+            rows.append(row)
+    header.insert(0, "Path")
+    if with_rev:
+        header.insert(0, "Revision")
```

Comment:
I think there should be a cleaner way to do this. E.g. no need to flatten the same metrics twice, etc. Give it a look, see if you can clean up this code a bit more.

Reply:
Any suggestions for this @efiop? The first flattening is required to get the header names

Reply:
There is no reason to flatten it twice though. You could flatten and then reuse both.